### PR TITLE
Wrong behavior deleting alias

### PR DESCRIPTION
**Pull request by olcbean**
Indices wildcards were resolved against all indices and aliases. And if an alias matched, then all indices with this alias were returned as matching.

In other words
```
POST /_aliases
{
    "actions" : [
        { "add" : { "index" : "foo_foo", "alias" : "foo" } },
        { "add" : { "index" : "bar_bar", "alias" : "foo" } }
    ]
}
```
```
POST /_aliases 
{
  "actions": [
    {
      "remove": {
        "index": "foo*",
        "alias": "foo"
      }
    }
  ]
}
```
will actually return both indices `foo_foo` and `bar_bar` as the common alias `foo` matched the wildcard. Which led to deleting the alias for both indices.

Related to #23960

Relates to #10106

### CODE REVIEW

**Comment by jimczi** [comment]
This looks correct to me but I am not very familiar with aliases. 
@martijnvg can you take a look ?

File: core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java [hunk 1]
```diff
@@ -125,6 +126,47 @@ public void testAll() {
         assertThat(newHashSet(resolver.resolve(context, Arrays.asList("_all"))), equalTo(newHashSet("testXXX", "testXYY", "testYYY")));
     }
 
+    // issue #23960
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java]: In general, I would prefer a straightforward comment explaining the issue being tested. A link to a GitHub issues means that I have to be online and context switch to obtain more information.

File: core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java [hunk 2]
```diff
@@ -643,6 +643,33 @@ public void testConcreteIndicesWildcardWithNegation() {
         assertEquals(0, indexNames.length);
     }
 
+    // issue #23960
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java]: In general, I would prefer a straightforward comment explaining the issue being tested. A link to a GitHub issues means that I have to be online and context switch to obtain more information.

File: core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java (hunk 1 above)
Comment (olcbean) [inline core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java]: @jasontedor Good point! I just followed the lead of another test in this class which is abbreviated only with the issue number. 

May I assume that generally a test should be commented with a short description why the test has been introduced (and maybe the issue for further reference)?

Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java]: If the test and its purpose is straightforward, I would say that no comment is necessary. If the test is for an tricky bug, or its purpose is unclear, I would say a comment is a necessity.

Comment (olcbean) [inline core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java]: @jasontedor Thank you for answering a trivial question! I understand this is common sense, but somehow common sense manages to differ between projects... Do you know if there is a resource where the basic coding practices are defined for new comers? If there is, a link will be really appreciated.

Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java]: I think a lot of what you're looking for is not written down, it's tribal knowledge that we all accumulate with experience in the project. We explain some straightforward things in the contributing docs but the vast majority of our conventions are not written down (and I think that's okay, we catch them during code review).

**Comment by nik9000** [comment]
Also these things change with time. Now we prefer comments explaining the
test without a link because that is easier to read in line and if we need
the issue we can blame it. But this hasn't always been the norm.

Still, it'd be good to stick it in contributing.md.

**Comment by olcbean** [comment]
The index wildcard is usually resolved against all indices and aliases in the cluster. Only for the index alias action the index wildcard will be resolved only against the indices. 

I could have introduced a new `IndexOption` for this but as this is a corner case I opted for an isolated change.

**Comment by javanna** [comment]
hi @olcbean thanks for the PR! I think the same fix would be needed in get alias api (`TransportGetAliasesAction`) too. Also, the ability the expand wildcards only to indices (leaving aliases out) was discussed before in https://github.com/elastic/elasticsearch/issues/2318#issuecomment-109967149 so it could be useful in other places. I would go with adding an index option at this point. It may be a bit more complicated to do but I think after all it would be cleaner. Would you be up for this?

**Comment by javanna** [comment]
To be clear, I would make it an internal only new index option, that is not settable through REST, similar to `allowAliasesToMultipleIndices()` maybe called `ignoreAliases()` or something along those lines. All of the existing constants would continue with the current behaviour, but a new constant used only in these couple of apis would set this flag to `true`. Such a change needs to be made in a backwards compatible manner though, as the index options are serialized over the wire.

**Comment by olcbean** [comment]
@javanna thank you for your feedback and putting this into prospective :)
I just introduced a new `IndicesOption` ( `ignoreAliases` ) as you suggested.

However I am not really sure that with this change the #2318 can be solved directly. According to a later comment in the very same thread https://github.com/elastic/elasticsearch/issues/2318#issuecomment-265983368 

> expand wildcards to aliases and concrete indices. If any aliases are present, throw an exception

but with the new option all wildcards will be resolved only to indices, so no exception can possibly be thrown. Do you think that this is an acceptable deviation from the described behavior?

**Comment by javanna** [comment]
> Do you think that this is an acceptable deviation from the described behavior?

Yes, the reason why we went that route is that I was against introducing a new index option at the time, but given that we need in it in more than in one place now, I changed my mind :)

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java [hunk 3]
```diff
@@ -111,6 +112,13 @@ public boolean allowAliasesToMultipleIndices() {
...
 
+    /**
+     * @return whether aliases should be ignored (when resolving a wildcard)
+     */
+    public boolean ignoreAliases() {
+        return (id & IGNORE_ALIASES) != 0;
+    }
+    
     public void writeIndicesOptions(StreamOutput out) throws IOException {
         out.write(id);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: this method and the read need to handle backwards compatibility: see https://github.com/elastic/elasticsearch/commit/7548b2edb782a2732aca5e9bae9016c6a01cb6e6 for a similar change with bw comp layer from when we added `allowAliasesToMultipleIndices`

File: core/src/main/java/org/elasticsearch/action/admin/indices/alias/IndicesAliasesRequest.java [hunk 4]
```diff
@@ -61,7 +61,7 @@
 
     //indices options that require every specified index to exist, expand wildcards only to open indices and
     //don't allow that no indices are resolved from wildcard expressions
-    private static final IndicesOptions INDICES_OPTIONS = IndicesOptions.fromOptions(false, false, true, false);
+    private static final IndicesOptions INDICES_OPTIONS = IndicesOptions.fromOptions(false, false, true, false, true, false, true);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/admin/indices/alias/IndicesAliasesRequest.java]: I think that we have to change also the default value for GetAliasesRequest ?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 5]
```diff
@@ -677,31 +678,43 @@ private static IndexNotFoundException infe(String expression) {
...
             String fromPrefix = expression.substring(0, expression.length() - 1);
             char[] toPrefixCharArr = fromPrefix.toCharArray();
             toPrefixCharArr[toPrefixCharArr.length - 1]++;
             String toPrefix = new String(toPrefixCharArr);
-            return metaData.getAliasAndIndexLookup().subMap(fromPrefix, toPrefix);
+            SortedMap<String,AliasOrIndex> subMap = metaData.getAliasAndIndexLookup().subMap(fromPrefix, toPrefix);
+            if (context.getOptions().ignoreAliases()) {
+                 return subMap.entrySet().stream()
+                        .filter(p -> !p.getValue().isAlias())
+                        .collect(Collectors.toMap(p -> p.getKey(), p -> p.getValue()));
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: nit: replace the lambda expressions with method references? `Map.Entry::getKey` and `Map.Entry::getValue` ?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 6]
```diff
@@ -677,31 +678,43 @@ private static IndexNotFoundException infe(String expression) {
...
+            return subMap;
         }
 
-        private static Map<String, AliasOrIndex> otherWildcard(MetaData metaData, String expression) {
+        private static Map<String, AliasOrIndex> otherWildcard(Context context, MetaData metaData, String expression) {
             final String pattern = expression;
             return metaData.getAliasAndIndexLookup()
                 .entrySet()
                 .stream()
+                .filter(e -> context.getOptions().ignoreAliases() ? !e.getValue().isAlias() : true)
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: `.filter(e -> context.getOptions().ignoreAliases() == false || e.getValue().isAlias() == false)`  ?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 7]
```diff
@@ -625,7 +626,7 @@ boolean isPreserveAliases() {
                 }
 
                 final IndexMetaData.State excludeState = excludeState(options);
-                final Map<String, AliasOrIndex> matches = matches(metaData, expression);
+                final Map<String, AliasOrIndex> matches = matches(context, metaData, expression);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: here we are fixing wildcard matching, but shouldn't we also honour the new option when resolving provided concrete names? e.g. what if I specify an alias amongst the indices, that are supposed to be concrete indices names only? I think if we fix this, it would also help with #10106 as we would not allow to create an alias that points to another alias anymore (which does not do what users think it does).

File: core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java [hunk 8]
```diff
@@ -643,6 +644,51 @@ public void testConcreteIndicesWildcardWithNegation() {
         assertEquals(0, indexNames.length);
     }
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java]: can you also add tests for the new option when resolving expressions that don't contain wildcards?

File: core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java [hunk 9]
```diff
@@ -643,6 +644,51 @@ public void testConcreteIndicesWildcardWithNegation() {
...
 
+    // concreteIndexNames must be able to resolve the provided wildcard only against the defined
+    // indices when ignoreAliases option is set or against the defined indices and aliases otherwise
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java]: I tend to think that this comment is not needed. Maybe add a small comment close to where you create the two indices options?

File: core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java [hunk 10]
```diff
@@ -125,6 +126,50 @@ public void testAll() {
         assertThat(newHashSet(resolver.resolve(context, Arrays.asList("_all"))), equalTo(newHashSet("testXXX", "testXYY", "testYYY")));
     }
 
+    // WildcardExpressionResolver must be able to resolve the provided wildcard only against the
+    // defined indices when ignoreAliases option is set or against the defined indices and aliases
+    // otherwise
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java]: same as above.

File: core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java [hunk 11]
```diff
@@ -125,6 +126,50 @@ public void testAll() {
...
+                IndexNameExpressionResolver.WildcardExpressionResolver
+                        .matches(context, state.getMetaData(), "f*o").keySet(),
+                equalTo(newHashSet("foo", "foo_foo")));
+        assertThat(
+                IndexNameExpressionResolver.WildcardExpressionResolver
+                        .matches(contextIndices, state.getMetaData(), "f*o").keySet(),
+                equalTo(newHashSet("foo_foo")));
+
+    }
+
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java]: would you mind adding a couple of integration tests to `IndexAliasesIT` also? Maybe test both IndicesAliasesRequest and GetAliasesRequest ?

**Review by javanna — CHANGES_REQUESTED** [review]
thanks for the PR @olcbean . I reviewed it and left some comments

File: core/src/main/java/org/elasticsearch/action/admin/indices/alias/IndicesAliasesRequest.java (hunk 4 above)
Comment (olcbean) [inline core/src/main/java/org/elasticsearch/action/admin/indices/alias/IndicesAliasesRequest.java]: Oh.. I was considering to leave the changes to `GetAliasesRequest` for another PR ( for better traceability ). But np, I will include them in this one.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/admin/indices/alias/IndicesAliasesRequest.java]: I am fine either way, if you prefer you can send another followup PR to address that.

**Comment by olcbean** [comment]
@javanna thank you for the review!

I hope the latest commit addresses the raised concerns ;)

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java [hunk 12]
```diff
@@ -120,12 +121,16 @@ public boolean ignoreAliases() {
     }
     
     public void writeIndicesOptions(StreamOutput out) throws IOException {
-        out.write(id);
+        if (out.getVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: not sure why this is still mark unreleased, we have already released alpha1, I think you should use alpha2.

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java [hunk 13]
```diff
@@ -111,13 +113,24 @@ public boolean allowAliasesToMultipleIndices() {
...
+    public boolean ignoreAliases() {
+        return (id & IGNORE_ALIASES) != 0;
+    }
+    
     public void writeIndicesOptions(StreamOutput out) throws IOException {
-        out.write(id);
+        if (out.getVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
+            out.write(id);
+        } else {
+            out.write(id & 0x3f);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: can you add a comment on what this does and why for future reference?

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java [hunk 14]
```diff
@@ -111,13 +113,24 @@ public boolean allowAliasesToMultipleIndices() {
...
+    /**
+     * @return whether aliases should be ignored (when resolving a wildcard)
+     */
+    public boolean ignoreAliases() {
+        return (id & IGNORE_ALIASES) != 0;
+    }
+    
     public void writeIndicesOptions(StreamOutput out) throws IOException {
-        out.write(id);
+        if (out.getVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: can you expand IndicesOptionsTests#testSerialization to pass in also the ignoreAliases flag when creating the random indices options and check that they get written/read correctly. You will need an if based on the version there as well.

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java [hunk 15]
```diff
@@ -133,8 +146,16 @@ public static IndicesOptions fromOptions(boolean ignoreUnavailable, boolean allo
...
 
-    static IndicesOptions fromOptions(boolean ignoreUnavailable, boolean allowNoIndices, boolean expandToOpenIndices, boolean expandToClosedIndices, boolean allowAliasesToMultipleIndices, boolean forbidClosedIndices) {
-        byte id = toByte(ignoreUnavailable, allowNoIndices, expandToOpenIndices, expandToClosedIndices, allowAliasesToMultipleIndices, forbidClosedIndices);
+    public static IndicesOptions fromOptions(boolean ignoreUnavailable, boolean allowNoIndices, boolean expandToOpenIndices,
+            boolean expandToClosedIndices, boolean allowAliasesToMultipleIndices, boolean forbidClosedIndices) {
+        return fromOptions(ignoreUnavailable, allowNoIndices, expandToOpenIndices, expandToClosedIndices, allowAliasesToMultipleIndices,
+                forbidClosedIndices, false);
+    }
+
+    public static IndicesOptions fromOptions(boolean ignoreUnavailable, boolean allowNoIndices, boolean expandToOpenIndices,
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: can you expand IndicesOptionsTests#testFromOptions using and testing this new flag too?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 16]
```diff
@@ -677,31 +681,43 @@ private static IndexNotFoundException infe(String expression) {
...
             assert expression.length() >= 2 : "expression [" + expression + "] should have at least a length of 2";
             String fromPrefix = expression.substring(0, expression.length() - 1);
             char[] toPrefixCharArr = fromPrefix.toCharArray();
             toPrefixCharArr[toPrefixCharArr.length - 1]++;
             String toPrefix = new String(toPrefixCharArr);
-            return metaData.getAliasAndIndexLookup().subMap(fromPrefix, toPrefix);
+            SortedMap<String,AliasOrIndex> subMap = metaData.getAliasAndIndexLookup().subMap(fromPrefix, toPrefix);
+            if (context.getOptions().ignoreAliases()) {
+                 return subMap.entrySet().stream()
+                        .filter(p -> !p.getValue().isAlias())
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: can you do here as well `p.getValue().isAlias() == false` ?

File: core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java [hunk 17]
```diff
@@ -427,6 +428,23 @@ public void testDeleteAliases() throws Exception {
...
+
+        logger.info("--> adding [foo] alias to [foo_foo] and [bar_bar]");
+        assertAcked(admin().indices().prepareAliases().addAlias("foo_foo", "foo"));
+        assertAcked(admin().indices().prepareAliases().addAlias("bar_bar", "foo"));
+
+        assertAcked(admin().indices().prepareAliases().addAliasAction(AliasActions.remove().index("foo*").alias("foo")).execute().get());
+
+        assertThat(admin().indices().prepareAliasesExist("foo").get().exists(), equalTo(true));
+        assertThat(admin().indices().prepareAliasesExist("foo").setIndices("foo_foo").get().exists(), equalTo(false));
+        assertThat(admin().indices().prepareAliasesExist("foo").setIndices("bar_bar").get().exists(), equalTo(true));
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java]: nit: can you use assertEquals rather than `assertThat(..., equalTo())` ? I know we do that in quite some places in our codebase but we seem to be preferring the former over the latter nowadays.

File: core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java [hunk 18]
```diff
@@ -427,6 +428,23 @@ public void testDeleteAliases() throws Exception {
...
+        logger.info("--> creating index [foo_foo] and [bar_bar]");
+        assertAcked(prepareCreate("foo_foo"));
+        assertAcked(prepareCreate("bar_bar"));
+        ensureGreen();
+
+        logger.info("--> adding [foo] alias to [foo_foo] and [bar_bar]");
+        assertAcked(admin().indices().prepareAliases().addAlias("foo_foo", "foo"));
+        assertAcked(admin().indices().prepareAliases().addAlias("bar_bar", "foo"));
+
+        assertAcked(admin().indices().prepareAliases().addAliasAction(AliasActions.remove().index("foo*").alias("foo")).execute().get());
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java]: maybe adding a get aliases request to the mix as well before the deletion so we make sure that it resolves things correctly?

File: core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java [hunk 19]
```diff
@@ -643,6 +644,66 @@ public void testConcreteIndicesWildcardWithNegation() {
...
+        assertEquals(1, indexNamesIndexWildcard.length);
+        assertEquals("foo_foo", indexNamesIndexWildcard[0]);
+
+        indexNamesIndexWildcard = indexNameExpressionResolver.concreteIndexNames(state, ignoreAliasesOptions, "f*o");
+
+        assertEquals(1, indexNamesIndexWildcard.length);
+        assertEquals("foo_foo", indexNamesIndexWildcard[0]);
+
+        try {
+            indexNameExpressionResolver.concreteIndexNames(state, ignoreAliasesOptions, "foo");
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java]: you can use expectThrows here and do it in a single line.

File: core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java [hunk 20]
```diff
@@ -643,6 +644,66 @@ public void testConcreteIndicesWildcardWithNegation() {
...
+        assertTrue(indexNames.contains("bar_bar"));
+
+        indexNames = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(state, indicesAndAliasesOptions, "f*o"));
+
+        assertEquals(2, indexNames.size());
+        assertTrue(indexNames.contains("foo_foo"));
+        assertTrue(indexNames.contains("bar_bar"));
+
+        indexNames = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(state, indicesAndAliasesOptions, "foo"));
+
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java]: can you remove the empty lines between assigning indexNames the corresponding assertions? Also above? If you want to divide this into blocks you could also do:

```
{
    List<String> indexNames = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(state, indicesAndAliasesOptions, "foo*"));
    assert.....
}
```

This way you declare the list each time and it doesn't get reused.

File: core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java [hunk 21]
```diff
@@ -125,6 +126,59 @@ public void testAll() {
...
+    public void testConcreteIndicesWildcardAndAliases() {
+        MetaData.Builder mdBuilder = MetaData.builder()
+                .put(indexBuilder("foo_foo").state(State.OPEN).putAlias(AliasMetaData.builder("foo")))
+                .put(indexBuilder("bar_bar").state(State.OPEN).putAlias(AliasMetaData.builder("foo")));
+        ClusterState state = ClusterState.builder(new ClusterName("_name")).metaData(mdBuilder).build();
+
+        // when ignoreAliases option is not set, WildcardExpressionResolver resolves the provided
+        // expressions against the defined indices and aliases
+        IndicesOptions indicesAndAliasesOptions = IndicesOptions.fromOptions(false, false, true, false, true, false, false);
+        IndexNameExpressionResolver.Context context = new IndexNameExpressionResolver.Context(state, indicesAndAliasesOptions);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java]: could you rename this to highlight what it holds? contextIndicesAndAliases?

File: core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java [hunk 22]
```diff
@@ -125,6 +126,59 @@ public void testAll() {
...
+
+        // ignoreAliases option is set, WildcardExpressionResolver resolves the provided expressions
+        // only against the defined indices
+        IndicesOptions ignoreAliasesOptions = IndicesOptions.fromOptions(false, false, true, false, true, false, true);
+        IndexNameExpressionResolver.Context contextIndices = new IndexNameExpressionResolver.Context(state, ignoreAliasesOptions);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/WildcardExpressionResolverTests.java]: contextIndicesOnly ?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 23]
```diff
@@ -177,6 +178,9 @@ public IndexNameExpressionResolver(Settings settings) {
         final Set<Index> concreteIndices = new HashSet<>(expressions.size());
         for (String expression : expressions) {
             AliasOrIndex aliasOrIndex = metaData.getAliasAndIndexLookup().get(expression);
+            if (context.getOptions().ignoreAliases() && aliasOrIndex.isAlias()) {
+                aliasOrIndex = null;
+            }
             if (aliasOrIndex == null) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I'd change change this to `if (aliasOrIndex == null || (aliasOrIndex.isAlias() && context.getOptions().ignoreAliases()))` and remove the new if above

File: core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java [hunk 24]
```diff
@@ -427,6 +428,23 @@ public void testDeleteAliases() throws Exception {
...
+
+        assertThat(admin().indices().prepareAliasesExist("foo").get().exists(), equalTo(true));
+        assertThat(admin().indices().prepareAliasesExist("foo").setIndices("foo_foo").get().exists(), equalTo(false));
+        assertThat(admin().indices().prepareAliasesExist("foo").setIndices("bar_bar").get().exists(), equalTo(true));
+        expectThrows(IndexNotFoundException.class, () -> admin().indices().prepareAliases()
+                .addAliasAction(AliasActions.remove().index("foo").alias("foo")).execute().actionGet());
     }
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java]: can you also add a test that makes sure that when adding an alias, you can not make it point to another alias anymore as the indices get resolved to indices only?

**Review by javanna — CHANGES_REQUESTED** [review]
Hi @olcbean thanks a lot for updating, I did another review round, it looks very good, I left some more comments but nothing major, it's getting close!

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java (hunk 12 above)
Comment (olcbean) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: I will need to rebase to see alpha2 ;)

Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: oh good point, maybe leave that as the last change then so it doesn't mess up the diffs for review :)

**Comment by olcbean** [comment]
@javanna thank you for your remarks! I will integrate them later in the week. 

Just one more question : do I need to modify some documentation as well ( maybe [indices.asciidoc](https://github.com/elastic/elasticsearch/blob/master/docs/reference/migration/migrate_6_0/indices.asciidoc) )?

**Comment by javanna** [comment]
very good point @olcbean . A note on the migrate guide would be nice, although we are fixing a bug. We are introducing a change in the behaviour of get aliases and indices aliases api so better be on the safe side and list it there.

File: core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java (hunk 18 above)
Comment (olcbean) [inline core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java]: Unfortunately the `GetAliasesRequest` is not yet fixed ( will be done in a separate PR ) so I don't think that it would bring anything adding it here at this stage as it will resolve to both indices ...

But I do agree that it would be nice to add this check once the `GetAliasesrequest` is changed ;)

Comment (javanna) [inline core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java]: sorry I was almost sure I saw a change in GetAliasesRequest too :) I was wrong.

**Comment by olcbean** [comment]
@javanna thanks for the remarks. 

I just pushed a newer version addressing your remarks

**Review by javanna — APPROVED** [review]
thanks for all the rounds @olcbean this looks great. I left the last few cosmetic comments, I will merge this as soon as you addressed those.

As for follow-up changes, it was a good move to not change get aliases and aliases exists api. We have decided to rather remove the indices part from those apis. We'd still want to use this new index option in the delete index api as I mentioned during the review. Let me know if you are up for working on those changes too.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 25]
```diff
@@ -693,31 +694,43 @@ private static IndexNotFoundException infe(String expression) {
             return excludeState;
         }
 
-        private static Map<String, AliasOrIndex> matches(MetaData metaData, String expression) {
+        public static Map<String, AliasOrIndex> matches(Context context, MetaData metaData, String expression) {
```
Comment (olcbean) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I had to change it to public for the unit test : `WildcardExpressionResolverTests#testConcreteIndicesWildcardAndAliases` 

Of course I can remove the test and revert to `private`

File: core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java [hunk 26]
```diff
@@ -785,6 +803,21 @@ public void testCreateIndexWithAliasesFilterNotValid() {
...
+
+        logger.info("--> adding [week_20] alias to [2017-05-20]");
+        assertAcked(admin().indices().prepareAliases().addAlias("2017-05-20", "week_20"));
+
+        expectThrows(IndexNotFoundException.class, () -> admin().indices().prepareAliases()
+                .addAliasAction(AliasActions.add().index("week_20").alias("tmp")).execute().actionGet());
+
+        assertAcked(admin().indices().prepareAliases().addAliasAction(AliasActions.add().index("2017-05-20").alias("tmp")).execute().get());
+    }
+
```
Comment (olcbean) [inline core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java]: `IndexAliasesIT#testRemoveIndexAndReplaceWithAlias` kind of tests this. I could add a test which explicitly tests this though.

**Comment by olcbean** [comment]
Thanks for the remarks @javanna! Just let me know if you think I should make some more changes.

As for the delete index (#2318), get aliases and aliases exists API changes, I would like to give them a try. Are there open issues for the get / exist aliases API changes?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 25 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I hadn't seen that, sorry

File: core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java [hunk 27]
```diff
@@ -804,16 +802,17 @@ public void testCreateIndexWithAliasesFilterNotValid() {
     }
 
     // aliases can be added only to indices
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java]: can you remote this comment? I think that the method name explains it well enough?

File: core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java (hunk 26 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java]: yea I think it would be good to have a test that makes sure that the index expression specified in remove_index resolves against indices only, rather than aliases and indices. I don't think the current test does that.

**Comment by javanna** [comment]
>  Are there open issues for the get / exist aliases API changes?

Last minute update is that we are not sure whether we should change those APIs, we keep on changing our minds (sorry!). The problem with the change that I had proposed is that in order to retrieve aliases, you need to know in advance whether the name you are specifying is an alias or an index. When retrieving aliases, it is kinda convenient that the index part is also resolved against aliases names, as confusing as it sounds. We decided to leave it as-is for now. Ping me once you send the delete index PR, would love to review that one too!

**Comment by olcbean** [comment]
@javanna, I just pushed some more changes. Hopefully the last round ;)

**Comment by javanna** [comment]
Thanks a lot @olcbean for the great work, I just merged this.

File: core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java (hunk 17 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java]: > we seem to be preferring the former over the latter nowadays

I disagree with this, I push towards `assertThat` when given the opportunity. It gives better error messages and it's easier to get `expected` versus `actual` in the correct order (`assertEquals` is messed up very frequently).

Comment (javanna) [inline core/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java]: oh man, I used to agree with you till somebody somehow convinced me that we should do it with assertEquals because it's simpler. I am sure there are comments in the github history around this. I really do not mind either way after all. Next time I will leave things as they are.
